core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java

```diff
@@ -89,35 +107,188 @@ private synchronized void check() {
             headers = builder.build();
         }
 
-        lastResponse = webClient.execute(headers);
-        lastResponse.aggregate().handle((res, cause) -> {
+        try (ClientRequestContextCaptor reqCtxCaptor = Clients.newContextCaptor()) {
+            lastResponse = webClient.execute(headers);
+            final ClientRequestContext reqCtx = reqCtxCaptor.get();
+            lastResponse.subscribe(new HealthCheckResponseSubscriber(reqCtx, lastResponse),
+                                   reqCtx.eventLoop(), SubscriptionOption.WITH_POOLED_OBJECTS);
+        }
+    }
+
+    @Override
+    public synchronized CompletableFuture<?> closeAsync() {
+        if (lastResponse != null) {
+            if (!closed) {
+                closed = true;
+                lastResponse.abort();
+            }
+            return lastResponse.completionFuture().handle((unused1, unused2) -> null);
+        } else {
+            return CompletableFuture.completedFuture(null);
+        }
+    }
+
+    private final class ResponseTimeoutUpdater extends SimpleDecoratingHttpClient {
+        ResponseTimeoutUpdater(HttpClient delegate) {
+            super(delegate);
+        }
+
+        @Override
+        public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+            if (maxLongPollingSeconds > 0) {
+                final long responseTimeoutMillis = ctx.responseTimeoutMillis();
+                if (responseTimeoutMillis > 0) {
```

Comment:
Should extend itself be a no-op if the timeout is 0?

Reply:
Good point. No need for `if`.